### PR TITLE
Make jax optional

### LINKED ISSUE #1701: make jax and the pybamm JaxSolver optional
**Summary**
make the pybamm JaxSolver an optional install, and jax/jaxlib an optional dependency

 **Motivation**
We pin the jax/jaxlib versions since their api is changing too rapidly, but this can cause dependency conflicts with other packages using jax, see #1699. Users should be able to install pybamm without the jax solver to resolve this.

### LINKED ISSUE #1775: Jax now supports mac with m1 chip

### PR DESCRIPTION
**Pull request by priyanshuone6**
# Description

Make jax and jaxlib optional

Fixes #1701 
Fixes #1775 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by tinosulzer** [comment]
Looks good, just a few more tests that need to be skipped if jax is not installed

File: tests/unit/test_citations.py [hunk 1]
```diff
@@ -255,10 +255,7 @@ def test_solver_citations(self):
             pybamm.IDAKLUSolver()
             self.assertIn("Hindmarsh2005", citations._papers_to_cite)
 
-    @unittest.skipIf(
-        system() == "Windows" or (system() == "Darwin" and "ARM64" in version()),
-        "JAX not supported on windows or Mac M1",
-    )
+    @unittest.skipIf(not(importlib.util.find_spec("jax")), "requires jax")
```
Comment (tinosulzer) [inline tests/unit/test_citations.py:258]: create a `pybamm.have_jax()` function, for example in the jax solver, for this (see https://github.com/pybamm-team/PyBaMM/blob/989811b5633437ff69dd226d11053b2e36862a50/pybamm/solvers/idaklu_solver.py#L20)

**Comment by tinosulzer** [comment]
Also, Jax should be installed at least for the ubuntu tests

**Comment by tinosulzer** [comment]
We should also say somewhere (maybe in the readme) which jax version should be installed for this to work.
Or, maybe, have a `pybamm.install_jax()` function that automatically installs the correct version?

File: tests/unit/test_citations.py (hunk 1 above)
Comment (priyanshuone6) [reply tests/unit/test_citations.py]: Added the have_jax in util because it was causing circular imports

**Comment by tinosulzer** [comment]
Re-running all the tests with and without jax is overkill (adds a lot of time to the CI). I think using the ubuntu tests to test with jax and the windows tests to test without jax is good enough. Otherwise we could specifically run only the jax tests separately, but I don't think that's necessary

**Comment by priyanshuone6** [comment]
> Looks good, just a few more tests that need to be skipped if jax is not installed

@tinosulzer Which tests are not yet covered

**Comment by tinosulzer** [comment]
These two:
```
======================================================================
ERROR: test_sensitivities (unit.test_solvers.test_base_solver.TestBaseSolver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/PyBaMM/PyBaMM/tests/unit/test_solvers/test_base_solver.py", line 352, in test_sensitivities
    solver.set_up(model, inputs={'a': 0, 'b': 0})
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/solvers/base_solver.py", line 441, in set_up
    use_jacobian=False,
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/solvers/base_solver.py", line 251, in process
    func = pybamm.EvaluatorJax(func)
AttributeError: module 'pybamm' has no attribute 'EvaluatorJax'

======================================================================
ERROR: test_ida_roberts_klu_sensitivities (unit.test_solvers.test_idaklu_solver.TestIDAKLUSolver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/PyBaMM/PyBaMM/tests/unit/test_solvers/test_idaklu_solver.py", line 74, in test_ida_roberts_klu_sensitivities
    model, t_eval, inputs={"a": a_value},
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/solvers/base_solver.py", line 908, in solve
    self.set_up(model, ext_and_inputs_list[0], t_eval)
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/solvers/base_solver.py", line 441, in set_up
    use_jacobian=False,
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/solvers/base_solver.py", line 251, in process
    func = pybamm.EvaluatorJax(func)
AttributeError: module 'pybamm' has no attribute 'EvaluatorJax'
```

**Review by tinosulzer — COMMENTED** [review]
Looks good other than that docs error. Perhaps the `JaxSolver` class should exist and simply raise an error in `__init__` if jax is not installed (so that `JaxSolver()` errors but the documentation builds). Alternatively, the simpler option is to install jax when building the docs

**Comment by priyanshuone6** [comment]
Jax is used everywhere in `pybamm/solvers/jax_bdf_solver.py` and it is difficult to skip the entire module unless I add a bunch of if statements in all the methods. Can I instead remove `from .solvers.jax_bdf_solver import jax_bdf_integrate` from `pybamm.__init__`?

**Comment by tinosulzer** [comment]
Can you install jax for coverage?

**Comment by priyanshuone6** [comment]
How do I test `if not pybamm.have_jax()`

**Comment by tinosulzer** [comment]
Windows tests should cover the not-jax case

**Review by tinosulzer — APPROVED** [review]
Looks good now, thanks

**Pull request merged**